rfcs/0008-serialization/0008-serialization.md

```diff
@@ -2,22 +2,255 @@
 Number: "0008"
 Category: Standards Track
 Status: Proposal
-Author: Ian Yang
+Author: Boyu Yang
 Organization: Nervos Foundation
 Created: 2018-12-17
 ---
 
 # Serialization
 
-CKB use two major serialization format, CFB and JSON.
+CKB use two major serialization format, [Molecule][molecule] and [JSON][json].
 
-[CFB][cfb] (Canonical FlatBuffers) is a restricted variant of FlatBuffers for producing unequivocal transfer syntax. Since CFB generated binary is still valid FlatBuffers, any FlatBuffers reader can parse the serialized messages, but only CFB builder can serialize messages into valid binary.
-
-CFB is in the proposal stage, and is not ready yet. Now plain FlatBuffers is
-used to serialize P2P messages.
+[Molecule][molecule] is a canonicalization and zero-copy serialization format.
+[Molecule][molecule] is in the experimental stage, and we may make few minor changes.
 
 [JSON][json] is used in node RPC service via [JSON-RPC][jsonrpc].
 
-[cfb]: https://github.com/nervosnetwork/rfcs/pull/47
+## Molecule
+
+### Summary
+
+#### Fixed Size or Dynamic Size
+
+| Type | byte  | array | struct | vector  |  table  | option  |  union  |
+|------|-------|-------|--------|---------|---------|---------|---------|
+| Size | Fixed | Fixed | Fixed  | Dynamic | Dynamic | Dynamic | Dynamic |
+
+#### Memory Layout
+
+```
+|  Type  |                      Header                      |               Body                |
+|--------+--------------------------------------------------+-----------------------------------|
+| array  |                                                  |  item-0 |  item-1 | ... |  item-N |
+| struct |                                                  | filed-0 | field-1 | ... | field-N |
+| fixvec | items-count                                      |  item-0 |  item-1 | ... |  item-N |
+| dynvec | full-size | offset-0 | offset-1 | ... | offset-N |  item-0 |  item-1 | ... |  item-N |
+| table  | full-size | offset-0 | offset-1 | ... | offset-N | filed-0 | field-1 | ... | field-N |
+| option |                                                  | item or none (zero bytes)         |
+| union  | item-type-id                                     | item                              |
+```
+
+- All items in Header is 32 bit unsigned integers in little-endian
+
+### Primitive Type
+
+#### `byte`
+
+The `byte` is a byte.
+
+##### Examples
+
+`00` is a `byte`.
+
+### Composite Types
+
+### `array`
+
+The `array` is a fixed-size type: it has a fixed-size inner type and a fixed length.
+The size of an `array` is the size of inner type times the length.
+
+Serialize an `array` only need to serialize all items in it. No extra cost for `array` itself.
```

Comment:
There's no overhead to serialize an `array`, which stores all items consecutively, without extra space between two adjacent items.